serix/serix.go

```diff
@@ -0,0 +1,747 @@
+}
+
+// WithLengthPrefixType specifies LengthPrefixType.
+func (ts TypeSettings) WithLengthPrefixType(lpt LengthPrefixType) TypeSettings {
+	ts.lengthPrefixType = &lpt
+	return ts
+}
+
+// LengthPrefixType returns LengthPrefixType.
+func (ts TypeSettings) LengthPrefixType() (LengthPrefixType, bool) {
+	if ts.lengthPrefixType == nil {
+		return 0, false
+	}
+	return *ts.lengthPrefixType, true
+}
+
+// WithObjectType specifies the object type. It can be either uint8 or uint32 number.
+// The object type holds two meanings: the actual code (number) and the serializer.TypeDenotationType like uint8 or uint32.
+// serix uses object type to actually encode the number
+// and to know its serializer.TypeDenotationType to be able to decode it.
+func (ts TypeSettings) WithObjectType(t interface{}) TypeSettings {
+	ts.objectType = t
+	return ts
+}
+
+// ObjectType returns the object type as an uint8 or uint32 number.
+func (ts TypeSettings) ObjectType() interface{} {
+	return ts.objectType
+}
+
+//WithLexicalOrdering specifies whether the type must be lexically ordered during serialization.
+func (ts TypeSettings) WithLexicalOrdering(val bool) TypeSettings {
+	ts.lexicalOrdering = &val
+	return ts
+}
+
+// LexicalOrdering returns lexical ordering flag.
+func (ts TypeSettings) LexicalOrdering() (val bool, set bool) {
+	if ts.lexicalOrdering == nil {
+		return false, false
+	}
+	return *ts.lexicalOrdering, true
+}
+
+// WithArrayRules specifies serializer.ArrayRules.
+func (ts TypeSettings) WithArrayRules(rules *ArrayRules) TypeSettings {
+	ts.arrayRules = rules
+	return ts
+}
+
+// ArrayRules returns serializer.ArrayRules.
+func (ts TypeSettings) ArrayRules() *ArrayRules {
+	return ts.arrayRules
+}
+
+func (ts TypeSettings) merge(other TypeSettings) TypeSettings {
+	if ts.lengthPrefixType == nil {
+		ts.lengthPrefixType = other.lengthPrefixType
+	}
+	if ts.objectType == nil {
+		ts.objectType = other.objectType
+	}
+	if ts.lexicalOrdering == nil {
+		ts.lexicalOrdering = other.lexicalOrdering
+	}
+	if ts.arrayRules == nil {
+		ts.arrayRules = other.arrayRules
+	}
+	return ts
+}
+
+func (ts TypeSettings) toMode(opts *options) serializer.DeSerializationMode {
+	mode := opts.toMode()
+	lexicalOrdering, set := ts.LexicalOrdering()
+	if set && lexicalOrdering {
+		mode |= serializer.DeSeriModePerformLexicalOrdering
+	}
+	return mode
+}
+
+// Encode serializes the provided object obj into bytes.
+// serix traverses the object recursively and serializes everything based on the type.
+// If a type implements the custom Serializable interface serix delegates the serialization to that type.
+// During the encoding process serix also performs the validation if such option was provided.
+// Use the options list opts to customize the serialization behaviour.
+// To ensure deterministic serialization serix automatically applies lexical ordering for maps.
+func (api *API) Encode(ctx context.Context, obj interface{}, opts ...Option) ([]byte, error) {
```

Comment:
It would be useful if the context passed to the API is passed onto bytes/syntactical validation functions, this would allow to pass protocol parameters which are needed for certain syntactical checks.

Reply:
Done

Reply:
Great, thanks!